HashStore-java 1.1.0 Release

doulikecookiedough opened the pull request:
Summary of Changes:
- New `HashStoreClient.java` class which can be used to interact directly with a HashStore (and respective test class).
- New utility class in `filehashstore` - `FileHashStoreUtility`, which contains generic methods and methods shared between `FileHashStore` and other classes.
- Expanded functionality to call `storeObject` without a pid
- New Public API methods (`findObject`, `tagObject`, `verifyObject`) and updates to `HashStore` interface for overload methods.
- Reference system under the hood to manage and keep track of reference files to store, retrieve and delete objects. 
- Refactored existing `deleteObject(String pid)` to delete an object, its reference files and all associated metadata documents. Plus new overload method `deleteObject(String idType, String id)` to enable deletion of object, references files and metadata documents - or just the object itself
- Refactored existing `deleteMetadata(String pid)` to delete all associated metadata documents for the given pid
- Refactored `storeMetadata` to store metadata for a given `pid` and `formatId` in a directory formed by calculating the hash of the given `pid`, with the document name being the hash of the `formatId`.
- Refactored 'ObjectMetadata' class to include a `pid` in its constructor and updated `storeObject` methods with a pid
- Updated HashStore interface javadocs and README.md

Hi @taojing2002, when you have scope - can you please help perform a code review for `HashStore-java`?

doulikecookiedough commented:
Hi @artntek, can you please also perform a code review when you free up next?

artntek commented on src/main/java/org/dataone/hashstore/Client.java:
General usability feedback: in the help messaging (see below), there are a lot of terms that are opaque to the user who is relatively smart but not well-versed in this domain. Here are a selection that could benefit from more description, an example, or a reference url etc:
 
* `Algorithm to use in calculations` vs. `Algorithm of HashStore`?
* `Delete an object from a HashStore` vs. `Delete objs from a HashStore`?
* `Depth of HashStore` & `Width of HashStore`?
* `Metadata format_id/namespace` & `Default metadata namespace`?
* `Specify testing with knbvm` (scary - could the user do any damage with this?? Do they need a login?)
* `Store object to a HashStore` vs. `Flag to store objs to a HashStore`?
...etc. You get the idea :-) 

Or maybe as a first step, it's as simple as adding a URL to the documentation?

```console
 $ java -cp target/hashstore-1.0-SNAPSHOT.jar org.dataone.hashstore.Client -h
usage: CommandLineApp
 -algo,--objectalgo <arg>                      Algorithm to use in
                                               calculations.
 -ap,--storealgo <arg>                         Algorithm of HashStore.
 -checksum,--obj_checksum <arg>                Checksum of object.
 -checksum_algo,--obj_checksum_algo <arg>      Algorithm of checksum
                                               supplied.
 -chs,--createhashstore                        Create a HashStore.
 -deletemetadata,--client_deletemetadata       Delete a metadata obj from
                                               a HashStore.
 -deleteobject,--client_deleteobject           Delete an object from a
                                               HashStore.
 -dfs,--delfromhs                              Delete objs from a
                                               HashStore.
 -dp,--storedepth <arg>                        Depth of HashStore.
 -format_id,--metadata_format <arg>            Metadata
                                               format_id/namespace
 -getchecksum,--client_getchecksum             Get the hex digest of a
                                               data object in a HashStore
 -h,--help                                     Show help options.
 -knbvm,--knbvmtestadc                         Specify testing with knbvm.
 -nobj,--numberofobj                           Number of objects to work
                                               with.
 -nsp,--storenamespace <arg>                   Default metadata namespace
 -path,--filepath <arg>                        Path to object.
 -pid,--pidguid <arg>                          PID or GUID of object.
 -rav,--retandval                              Retrieve and validate objs
                                               from a HashStore.
 -retrievemetadata,--client_retrievemetadata   Retrieve a metadata obj
                                               from a HashStore.
 -retrieveobject,--client_retrieveobject       Retrieve an object from a
                                               HashStore.
 -sdir,--storedirectory <arg>                  Location of objects to
                                               convert.
 -size,--obj_size <arg>                        Size of object
 -store,--storepath <arg>                      Path to HashStore.
 -storemetadata,--client_storemetadata         Store metadata to a
                                               HashStore
 -storeobject,--client_storeobject             Store object to a
                                               HashStore.
 -sts,--storetohs                              Flag to store objs to a
                                               HashStore
 -stype,--storetype <arg>                      Type of store 'objects' or
                                               'metadata'
 -wp,--storewidth <arg>                        Width of HashStore.
```

doulikecookiedough replied:
I totally see where you're coming from and I've updated the descriptions. I think (the below) is an improvement, and will continue to revise.

```console
doumok@Dous-MacBook-Pro hashstore-java % java -cp target/hashstore-1.0-SNAPSHOT.jar org.dataone.hashstore.Client -h
usage: HashStore Client
 -algo,--objectalgo <arg>                      Algorithm to use when
                                               calling '-getchecksum' or
                                               '-storeobject' flag.
 -ap,--storealgo <arg>                         Algorithm used for
                                               calculating file addresses
                                               in a HashStore.
 -checksum,--obj_checksum <arg>                Checksum of object to
                                               store.
 -checksum_algo,--obj_checksum_algo <arg>      Algorithm of checksum
                                               supplied.
 -chs,--createhashstore                        Flag to create a HashStore.
 -deletemetadata,--client_deletemetadata       Flag to delete metadata
                                               objs from a HashStore.
 -deleteobject,--client_deleteobject           Flag to delete objs from a
                                               HashStore.
 -dfs,--delfromhs                              (knbvm) Test flag to delete
                                               objs from a HashStore
 -dp,--storedepth <arg>                        Depth of HashStore to
                                               create.
 -format_id,--metadata_format <arg>            Format_id/namespace of
                                               metadata to store, retrieve
                                               or delete.
 -getchecksum,--client_getchecksum             Flag to get the hex digest
                                               of a data object in a
                                               HashStore.
 -h,--help                                     Show help options.
 -knbvm,--knbvmtestadc                         (knbvm) Flag to specify
                                               testing with knbvm.
 -nobj,--numberofobj                           (knbvm) Option to specify
                                               number of objects to
                                               retrieve from a Metacat db.
 -nsp,--storenamespace <arg>                   Default metadata namespace
                                               in a HashStore.
 -path,--filepath <arg>                        Path to object/metadata.
 -pid,--pidguid <arg>                          PID or GUID of
                                               object/metadata.
 -rav,--retandval                              (knbvm) Test flag to
                                               retrieve and validate objs
                                               from a HashStore.
 -retrievemetadata,--client_retrievemetadata   Flag to retrieve metadata
                                               objs from a HashStore.
 -retrieveobject,--client_retrieveobject       Flag to retrieve objs from
                                               a HashStore.
 -sdir,--storedirectory <arg>                  (knbvm) Option to specify
                                               the directory of objects to
                                               convert.
 -size,--obj_size <arg>                        Size of object to
                                               store/validate.
 -store,--storepath <arg>                      Path to HashStore.
 -storemetadata,--client_storemetadata         Flag to store metadata to a
                                               HashStore
 -storeobject,--client_storeobject             Flag to store objs to a
                                               HashStore.
 -sts,--storetohs                              (knbvm) Test flag to store
                                               objs to a HashStore
 -stype,--storetype <arg>                      (knbvm) Option to specify
                                               'objects' or 'metadata'
 -wp,--storewidth <arg>                        Width of HashStore to
                                               create.
```

doulikecookiedough commented on README.md:
```diff
@@ -24,6 +24,45 @@ and then install or build the package with `mvn install` or `mvn package`, respe
+$ mvn clean package -Dmaven.test.skip=true
+
+# Get help
+$ java -cp /path/to/hashstore-1.0-SNAPSHOT.jar org.dataone.hashstore.Client -h
```
Thank you for the suggestion - that makes sense! I have updated the README.md

doulikecookiedough commented on pom.xml:
```diff
@@ -25,6 +24,11 @@
+    </dependency>
     <dependency>
       <groupId>junit</groupId>
       <artifactId>junit</artifactId>
```
That makes sense - I have upgraded to JUnit5 and updated all test classes.

doulikecookiedough commented on src/main/java/org/dataone/hashstore/HashStore.java:
Understood - thank you! I've updated the interface method for the throws section in the affected methods - as well as the javadocs to be simpler.

doulikecookiedough commented on src/main/java/org/dataone/hashstore/HashStore.java:
```diff
@@ -59,6 +59,18 @@ ObjectInfo storeObject(
                 String checksumAlgorithm, long objSize
         ) throws NoSuchAlgorithmException, IOException, PidObjectExistsException, RuntimeException;
 
+        ObjectInfo storeObject(
+                InputStream object, String pid, String checksum, String checksumAlgorithm
+        ) throws NoSuchAlgorithmException, IOException, PidObjectExistsException, RuntimeException;
+
+        ObjectInfo storeObject(InputStream object, String pid, String additionalAlgorithm)
+                throws NoSuchAlgorithmException, IOException, PidObjectExistsException,
+                RuntimeException;
+
+        ObjectInfo storeObject(InputStream object, String pid, long objSize)
+                throws NoSuchAlgorithmException, IOException, PidObjectExistsException,
+                RuntimeException;
+
```
I like that! Javadocs have been added.

doulikecookiedough commented on src/main/java/org/dataone/hashstore/HashStore.java:
```diff
@@ -84,6 +96,10 @@ String storeMetadata(InputStream metadata, String pid, String formatId) throws I
                 IllegalArgumentException, FileNotFoundException, InterruptedException,
                 NoSuchAlgorithmException;
 
+        String storeMetadata(InputStream metadata, String pid) throws IOException,
```
Javadocs have been added.

doulikecookiedough commented on src/main/java/org/dataone/hashstore/Client.java:
```diff
@@ -0,0 +1,862 @@
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
+
+public class Client {
```
I've added the missing JavaDoc - apologies, in hindsight - I can't remember why I thought it was okay to omit it here (and in the main method feedback below).

doulikecookiedough commented on src/main/java/org/dataone/hashstore/Client.java:
```diff
@@ -0,0 +1,862 @@
+                    int bytesRead = objStream.read(buffer, 0, buffer.length);
+                    String objPreview = new String(buffer, 0, bytesRead, StandardCharsets.UTF_8);
+                    objStream.close();
+                    System.out.println(objPreview);
```
I've added the following string to print afterwards:
- "...\n<-- Truncated for Display Purposes -->"

doulikecookiedough commented on src/main/java/org/dataone/hashstore/filehashstore/FileHashStoreUtility.java:
```diff
@@ -0,0 +1,117 @@
+     * @param argument Value that is being checked
+     * @param method   Calling method or class
+     */
+    public static void ensureNotNull(Object object, String argument, String method) {
```
I agree that it does feel more informative to throw an `IllegalArgumentException` exception here instead - I've updated the method. Thank you for the suggestion!

doulikecookiedough commented on src/main/java/org/dataone/hashstore/filehashstore/FileHashStoreUtility.java:
```diff
@@ -0,0 +1,117 @@
+     */
+    public static void checkForEmptyString(String string, String argument, String method) {
+        if (string.trim().isEmpty()) {
+            String errMsg = "FileHashStoreUtility.isStringNullOrEmpty - Calling Method: " + method
```
Apologies for the confusion - previously I combined the check for checking nulls and for an empty string, but have separated it since and missed updating the exception messaging here (and in your other comments).

doulikecookiedough commented on src/main/java/org/dataone/hashstore/filehashstore/FileHashStoreUtility.java:
```diff
@@ -0,0 +1,117 @@
+     * @param object Object to check
+     * @param method Calling method
+     */
+    public static void checkNotNegative(long object, String method) {
```
Understood, updated and thank you!

doulikecookiedough commented on src/main/java/org/dataone/hashstore/filehashstore/FileHashStoreUtility.java:
```diff
@@ -0,0 +1,117 @@
+     * @param method   Calling method
+     */
+    public static void checkForEmptyString(String string, String argument, String method) {
+        if (string.trim().isEmpty()) {
```
I've called `ensureNotNull()` to check for NPE here before evaluating the string.

doulikecookiedough commented:
Thank you for your feedback and code review @artntek! I have incorporated most if not all of your suggestions and resolved our conversations. If you have any other comments, please let me know 🙏 .

artntek commented on src/test/java/org/dataone/hashstore/ClientTest.java:
```diff
@@ -15,12 +10,12 @@
 import java.util.List;
 import java.util.Properties;
 
+import static org.junit.jupiter.api.Assertions.*;
```
best practice to call out imports explicitly, rather than use `.*`

doulikecookiedough replied:
Noted - thank you!

artntek commented on src/test/java/org/dataone/hashstore/HashStoreTest.java:
```diff
@@ -1,22 +1,18 @@
 import java.nio.file.Path;
 import java.util.Properties;
 
+import static org.junit.jupiter.api.Assertions.*;
```
best practice to call out imports explicitly, rather than use `.*`

artntek commented on src/test/java/org/dataone/hashstore/ObjectInfoTest.java:
```diff
@@ -3,11 +3,10 @@
 
-import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertNotNull;
+import static org.junit.jupiter.api.Assertions.*;
```
best practice to call out imports explicitly, rather than use `.*`

artntek commented on src/test/java/org/dataone/hashstore/filehashstore/FileHashStoreInterfaceTest.java:
```diff
@@ -24,12 +24,11 @@
+import org.junit.jupiter.api.io.TempDir;
 
-import static org.junit.Assert.*;
+import static org.junit.jupiter.api.Assertions.*;
```
best practice to call out imports explicitly, rather than use `.*`

artntek commented on src/test/java/org/dataone/hashstore/filehashstore/FileHashStoreProtectedTest.java:
```diff
@@ -14,18 +14,14 @@
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+
+import static org.junit.jupiter.api.Assertions.*;
```
best practice to call out imports explicitly, rather than use `.*`

artntek commented on src/test/java/org/dataone/hashstore/filehashstore/FileHashStorePublicTest.java:
```diff
@@ -9,15 +9,12 @@
-import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertTrue;
-import static org.junit.Assert.fail;
+import static org.junit.jupiter.api.Assertions.*;
```
best practice to call out imports explicitly, rather than use `.*`

doulikecookiedough commented on README.md:
```diff
@@ -32,35 +32,35 @@ We also maintain a parallel [Python-based version of HashStore](https://github.c
 
 # Get help
-$ java -cp /path/to/hashstore-1.0-SNAPSHOT.jar org.dataone.hashstore.Client -h
+$ java -cp ./target/.../hashstore-1.0-SNAPSHOT.jar org.dataone.hashstore.Client -h
```
Fixed!

artntek commented on src/test/java/org/dataone/hashstore/filehashstore/FileHashStoreInterfaceTest.java:
```diff
@@ -28,7 +35,8 @@
 import org.junit.jupiter.api.io.TempDir;
 
-import static org.junit.jupiter.api.Assertions.*;
+// import static org.junit.jupiter.api.Assertions.*;
```
commented code

taojing2002 commented on src/main/java/org/dataone/hashstore/Client.java:
```diff
@@ -0,0 +1,905 @@
+ * HashStore's 'Client' class is a command line interface that allows a developer to create a new
+ * HashStore or interact directly with an existing HashStore. See 'README.md' for usage examples.
+ */
+public class Client {
```
The name of the class - Client seems too general to me. Also is it really a client? It seems a tool class. May we rename it to something like - CommandLineTool?

doulikecookiedough replied:
No problem, it has been renamed.

mbjones replied:
It appears to still be named `Client.java` to me. I agree with @taojing2002 that a more descriptive class name would make sense. Even `CommandLineTool` is too generic. How about `HashStoreClient`?  If we had a shell script to invoke it (or the python version, it would be nice if it were short and mnemonic, like `hashstore`, or `hs` or `hscli`.

doulikecookiedough replied:
@mbjones Sorry for the confusion - I hadn't yet pushed my changes, but `HashStoreClient` sounds good to me! I will also create a new issue for hashstore to update the `client.py` module name as well.

taojing2002 commented on src/main/java/org/dataone/hashstore/filehashstore/FileHashStore.java:
You use  `String METADATA_NAMESPACE` to store the metadata namespace. So this means we can only store one metadata type. To my understanding, we can store multiple types.

taojing2002 replied:
I am not sure what the method `syncPubObject` mean. Do you mean `syncPutObject`?

taojing2002 replied:
The method of `deleteObjectAndParentDirectories` makes me nervous since it deletes the parent directory. This is method is not synchronized, so the checking empty process is not always right. It may cause some issues.

doulikecookiedough replied:
Ah, my inclination here is to synchronize the `deleteObject` and `deleteMetadata` method given your feedback - but may I ask for your thoughts on leaving empty folders (and not deleting them)? While it might be a little "messy" if someone were to physically navigate through the HashStore, it could simplify the deletion process.

mbjones replied:
If these are just the prefixed two digit directories, I think it is just fine to leave them.

doulikecookiedough replied:
Thank you for your feedback! Yes, they are the prefixed two digit directories. I will leave them and have updated the methods that call `deleteObjectAndParentDirectories`.

taojing2002 commented on src/main/java/org/dataone/hashstore/filehashstore/FileHashStore.java:
```diff
@@ -230,7 +229,6 @@ protected void verifyHashStoreProperties(
             );
 
```
line 238 should be "store metadata namespace"

doulikecookiedough replied:
Got it, thank you!

taojing2002 commented on src/main/java/org/dataone/hashstore/filehashstore/FileHashStore.java:
```diff
@@ -269,15 +267,13 @@ protected void verifyHashStoreProperties(
+            HashMap<?, ?> hashStoreYamlProperties = om.readValue(hashStoreYamlFile, HashMap.class);
             hsProperties.put(
                 HashStoreProperties.storeDepth.name(), hashStoreYamlProperties.get("store_depth")
             );
```
in the `writeHashStoreYaml` method, you need to close the writer.

doulikecookiedough replied:
Understood, it has been fixed.

taojing2002 commented on src/main/java/org/dataone/hashstore/filehashstore/FileHashStore.java:
In the storeObject methods, we can put the object size to be 0 to skip the size checking process. Do you have some specific reason to do that? If the metacat client puts the size to be 0 in the system metadata, the hash store will skip the checking, right? So they can bypass the quota service.

doulikecookiedough replied:
If the metacat client supplies size to be `0` when calling `storeObject`, it will ignore the object size check to verify the object size (but still store the object). I thought that checking for `> 0` would be a simple way to skip the check if desired.

May I ask if this is ok? Or do you think that if metacat client calls `storeObject` with an object size to validate, we shouldn't accept 0.

mbjones replied:
Why would we want an option to skip checking the object size? And what happens if a file truly of size 0 is added? Seems like the size check should be consistent and not skipped, unless there is some specific need we can identify.

doulikecookiedough replied:
To provide context, I added overloaded methods so that the metacat client could call `storeObject` to store and only validate the object size, to only generate an additional algorithm or to only validate a checksum and checksum algorithm if desired.
- If the overloaded method that only stores and validates an object size is called, there is only a check to ensure that the object size is not negative.
- The only other way to call `storeObject` with 0 as the object size is to use the comprehensive `storeObject` method (where you must provide values for all possible arguments), and this only checks that the size is not negative as well.

I had initially thought that this was a good way to manage the overloaded methods - but I understand now how I've created an inconsistent (and possibly unexpected) experience. I have refactored my overloaded methods, and instead of allowing 0 to be passed, I added a check for `-1` as the option to bypass the size validation. This feels much more intentional and less ambiguous to me, as we would never intentionally provide a negative number. 0 is also now never acceptable.

taojing2002 commented on src/main/java/org/dataone/hashstore/filehashstore/FileHashStore.java:
You uses a lot of method overloading in this class. At least some of them can just call the base method with default parameters. So this can eliminate some duplicate code.

doulikecookiedough replied:
Ah - I see it now, thank you!

taojing2002 commented on src/main/java/org/dataone/hashstore/filehashstore/FileHashStoreUtility.java:
```diff
@@ -0,0 +1,124 @@
+
+            }
+            // Close dataStream
+            dataStream.close();
```
Closing the data stream should be in the finally clause.

doulikecookiedough replied:
Understood, thank you!

taojing2002 commented on src/main/java/org/dataone/hashstore/filehashstore/FileHashStore.java:
On line 1129, may you use `equalsIgnoreCase` to replace `equals` when you compare two checksums?

doulikecookiedough replied:
Got it, thank you!

doulikecookiedough commented on src/main/java/org/dataone/hashstore/filehashstore/FileHashStore.java:
```diff
@@ -157,8 +158,7 @@ public FileHashStore(Properties hashstoreProperties) throws IllegalArgumentExcep
         Path hashstoreYaml = STORE_ROOT.resolve("hashstore.yaml");
```
I agree, and have made the changes. Thank you!

taojing2002 commented:
Synchronizing the deleting methods doesn't help. If we really want to
delete empty directories, we should somehow lock the file system, which is
hard to do. So keeping the empty directories may be the easiest and safest
way.

On Thu, Oct 19, 2023 at 7:20 PM Matt Jones ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> On src/main/java/org/dataone/hashstore/filehashstore/FileHashStore.java
> <https://github.com/DataONEorg/hashstore-java/pull/48#discussion_r1366366563>
> :
>
> If these are just the prefixed two digit directories, I think it is just
> fine to leave them.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/DataONEorg/hashstore-java/pull/48#discussion_r1366366563>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB5QQDHTL6KQJK6NJN4LBFLYAHNW3AVCNFSM6AAAAAA45LYXP6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTMOBZGA2TINRWG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

doulikecookiedough commented:
Thank you for clarifying and the feedback @taojing2002 (and @mbjones)! I believe I've addressed your feedback, and have pushed my latest commits to the develop branch - please let me know if you have any concerns with these changes (or any other feedback).

taojing2002 commented on src/main/java/org/dataone/hashstore/filehashstore/FileHashStore.java:
```diff
@@ -765,22 +757,23 @@ public InputStream retrieveMetadata(String pid, String formatId) throws Exceptio
     /**
      * Overload method for retrieveMetadata with default metadata namespace
      */
-    public InputStream retrieveMetadata(String pid) throws Exception {
```
Can this method call retrieveMetadata(pid, formatid)? So we can get rid of the duplicated code.

doulikecookiedough replied:
I have updated both overload methods, as well as cleaned up the logging statements. Thank you!

taojing2002 commented on src/main/java/org/dataone/hashstore/filehashstore/FileHashStore.java:
In the putObject method, you have two places to check if the file exists - `Files.exists(objRealPath))`. At the line 997, the code throws an exception if the duplicated file exist. But at line 1038, it just logs the info. Why the behaviors are different?

doulikecookiedough replied:
It is different because I am trying to account for the race conditions that are created when we call `storeObject` multiple times on the same pid (accidentally or not).
- The first time we call `Files.exists(objRealPath))` is before any temporary file is created. If the object already exists, we throw an exception immediately. 
- The second time `Files.exists(objRealPath))` is called is after we write the `InputStream object` to a temporary file. This happens right before we determine whether we should move it to its permanent address or not.

If multiple threads manage to reach this same point and bypass the synchronization, there will be 2+ temporary files. The quickest thread will move its temporary file first to the permanent location. So the other threads should see that it's already there (second time we call `Files.exists(objRealPath))`), and then proceed to delete its respective (and now redundant) temporary file.

taojing2002 commented on src/main/java/org/dataone/hashstore/filehashstore/FileHashStore.java:
In method `validateTmpObject`, at line 1087, the error message  should include the sizes just like you do on line 1093. 
At line 1122, the error message says the object can't be validate. Actually the error is that the checksums don't match. You may put the checksums in the error message as you do on line 1127.

doulikecookiedough replied:
Understood - I have updated the inaccurate logging statements.

doulikecookiedough commented:
The necessary changes to be integrated into the Metacat client have been made. However, after further discussions, we have determined that there is still one missing piece remaining to complete the 1.0.0 release described here: [Refactor `deleteObject` for new requirements](https://github.com/DataONEorg/hashstore-java/issues/57). 

Additionally, the recent changes have been tested with test data found in `knbvm (test.arcticdata.io)` with up to 100,000 objects which includes numerous persisten identifiers that reference the same object (`cid`).

doulikecookiedough committed:
Clean up 'HashStoreRunnable' and add/revise logging statements

doulikecookiedough committed:
Extract new protected method 'storeHashStoreRefsFiles' from 'tagObject'

doulikecookiedough committed:
Organize new 'storeHashStoreRefsFile' method in 'FileHashStore' class and update javadoc

doulikecookiedough committed:
Revise junit tests for 'tagObject', add new junit tests for 'storeHashStoreRefsFiles' and add todo items

doulikecookiedough committed:
Add new method 'unTagObject'

doulikecookiedough committed:
Revise comment for accuracy

doulikecookiedough committed:
Add new junit test for 'unTagObject' and fix bug in existing 'tagObject' tests when checking for number of files

doulikecookiedough committed:
Add new junit tests for 'unTagObject'

doulikecookiedough committed:
Revise 'findObject' to throw custom exception 'PidRefsFileNotFoundException' when it cannot find a pid refs file instead of generic 'FileNotFoundException'

doulikecookiedough committed:
Add new junit test for 'unTagObject' and revise 'findObject' junit test to assert newly added custom exception

doulikecookiedough committed:
Add clean-up code to 'unTagObject' when a pid refs file does not exist, but pid is tagged in the cid refs file, and add new junit test

doulikecookiedough committed:
Revise 'unTagObject' javadoc and comments

doulikecookiedough committed:
Add print statements to 'HashStoreRunnable' to assist with debugging inconsistent test failures

doulikecookiedough committed:
Add 'unTagObject' to 'tagObject' process

doulikecookiedough committed:
Add new junit tests for 'updateRefsFile'

doulikecookiedough committed:
Add new junit tests for 'tagObject' and fix bug where a pid refs is deleted when 'tagObject' is called to tag the pid that is already referencing another cid

doulikecookiedough committed:
Revise comments and add additional print statement to assist with debugging for 'deleteObject_1000Pids_1Obj_viaRunnable' junit test

doulikecookiedough committed:
Fix inaccurate argument string in 'verifyChecksumParameters'

doulikecookiedough committed:
Cleanup 'FileHashStoreReferencesTest' class

doulikecookiedough committed:
Merge pull request #89 from DataONEorg/feature-83-tagobject-cleanup

Feature-83: `tagObject` Refactor

doulikecookiedough committed:
Remove 'findObject' from 'HashStore' interface and move javadoc to 'FileHashStore'

doulikecookiedough committed:
Remove 'findObject' from 'HashStoreClient' and update README.md

doulikecookiedough committed:
Cleanup test classes for ide warnings and re-organize junit tests

doulikecookiedough committed:
Move 'findObject' code in 'FileHashStore' to core and supporting methods section

doulikecookiedough committed:
Merge pull request #90 from DataONEorg/feature-87-removefo-publicapi

Feature-87: Remove `findObject` from the Public API

doulikecookiedough committed:
Code clean-up: Review/revise comments, refactor 'retrieveMetadata' methods to be DRY by extracting method 'getMetadataDocInputStream'

doulikecookiedough committed:
Code clean-up: Review/revise comments, refactor 'deleteMetadata' methods to be DRY by extracting method 'syncRenameMetadataDocForDeletion'

doulikecookiedough committed:
Code clean-up: Revise deleteMetadata to call existing method to get path to metadata document, refactor 'isDefaultAlgorithm' into 'shouldCalculateAlgorithm' to improve clarity

doulikecookiedough committed:
Code clean-up: Remove unused 'deleteRefsFile' method & update junit tests, revise comments, resolve linting warnings

doulikecookiedough committed:
Resolve linting warning in 'putObject' by replacing .delete() call on tmpFile to using Files.delete()

doulikecookiedough committed:
Clean-up 'deleteObject' and 'unTagObject', refactor to begin with removing pid from cid refs file

doulikecookiedough committed:
Rename method 'getMetadataDocInputStream' to 'getHashStoreMetadataInputStream'

doulikecookiedough committed:
Code clean-up: remove redundant logging texts

doulikecookiedough committed:
Accept IDE warning suggestion to weaken specific list types to a Collection

doulikecookiedough committed:
Code clean-up: fix ide warnings for 'FileHashStoreUtility' class

doulikecookiedough committed:
Code clean-up: fix ide warnings for 'ObjectMetadata' class

doulikecookiedough committed:
Code clean-up: remove redundant printstacktrace statement in 'HashStoreFactory'

doulikecookiedough committed:
Code clean-up: address warnings for HashStore interface

doulikecookiedough committed:
Code clean-up: address warnings for HashStoreClient class

doulikecookiedough committed:
Code clean-up: Address additional warnigs from test classes

doulikecookiedough committed:
Accept ide warning to Weaken type from List to Collection in 'FileHashStore'

doulikecookiedough committed:
Refactor public api methods in 'FileHashStore' by extracting synchronization code

doulikecookiedough committed:
Remove unused parameter from 'syncRenameMetadataDocForDeletion' and fix affected code

doulikecookiedough committed:
Deactivate known intermittently failing test which issue has been created for

doulikecookiedough committed:
Update storeObject HashStoreRunnable junit test

doulikecookiedough committed:
Fix warning for 'verifyObject' with boolean value in signature by removing a data object by default if it is determined to be invalid and the right conditions are met

doulikecookiedough committed:
Remove todo item and add new junit tests for 'getHashStoreMetadataInputStream'

doulikecookiedough committed:
Add missing junit tests for 'verifyChecksumParameters'

doulikecookiedough committed:
Add missing junit tests for 'shouldCalculateAlgorithm'

doulikecookiedough committed:
Delete 'FileHashStoreReferencesTest' class, move and organize junit tests to 'FileHashStoreProtected', fix incorrect import in 'FileHashStoreInterface' class and use correct 'Disabled' decorator

doulikecookiedough committed:
Rename 'FileHashStorePublicTest' to 'FileHashStoreInitTest' for accuracy

doulikecookiedough committed:
Add missing junit tests for 'isStringInRefsFile' in 'FileHashStoreProtectedTest' class

doulikecookiedough committed:
Fix bug in 'validateTmpObject' and add missing 'validateTmpObject' junit tests

doulikecookiedough committed:
Reword javadoc for 'HashStoreClient' class

doulikecookiedough committed:
Re-activate disabled runnable test, update 'putObject' to delete redundant tmp file before logging warning

doulikecookiedough committed:
Rework/optimize 'putObject' routine and update junit tests

doulikecookiedough committed:
Revise error statement in 'putObject' and variable name in intermittently failing test to assist with debugging

doulikecookiedough committed:
Merge pull request #91 from DataONEorg/bug-88-tmpfile-cleanup

Bug-88: `tmpFile` Cleanup

doulikecookiedough committed:
Update README.md

doulikecookiedough committed:
Update HashStore interface, rename 'verifyObject' to 'deleteInvalidObject' in 'FileHashStore' and update junit tests

doulikecookiedough committed:
Merge pull request #93 from DataONEorg/feature-92-deleteinvalidobject

Feature-92: Rename `verifyObject` to `deleteInvalidObject`

doulikecookiedough committed:
Move synchronization code calls into try blocks

doulikecookiedough committed:
Add missing synchronization for data object cid when data object is about to be moved to its final location

doulikecookiedough committed:
Rename synchronization array from 'objectLockedIds' to 'objectLockedPids' to improve clarity

doulikecookiedough committed:
Rename synchronization array from 'metadataLockedIds' to 'metadataLockedDocIds' to improve clarity

doulikecookiedough committed:
Rename synchronization array from 'referenceLockedCids' to 'objectLockedCids' to improve clarity

doulikecookiedough committed:
Add new synchronization code for 'tagObject' which can be called directly, to synchronize on pid via 'referenceLockedPids'

doulikecookiedough committed:
Revise 'tagObject' unexpected exception process to release locks before calling 'unTagObject', which is synchronized, to prevent a deadlock situation

doulikecookiedough committed:
Check 'deleteInvalidObject' for when ObjectMetadata passed does not include hexDigests, add custom exception and two new junit tests

doulikecookiedough committed:
Rename FileHashStoreUtility 'checkForEmptyString' to 'checkForEmptyAndValidString' for clarity

doulikecookiedough committed:
Rework FileHashStoreUtility 'deleteListItems' method to attempt to delete as many items in the list as possible

doulikecookiedough committed:
Add logic to Public API store methods to try-finally .close the supplied stream to the object or metadata

doulikecookiedough committed:
Refactor 'deleteMetadata' with logic to revert the process if there is an unexpected issue, and add new junit tests for new supporting methods

doulikecookiedough committed:
Clean up code for recently reworked deleteMetadata process

doulikecookiedough committed:
Merge pull request #94 from DataONEorg/feature-57-delobjref-jingcrch

Feature-57: `deleteObject` Refactor Changes from Code Review

doulikecookiedough committed:
Revert process to restore metadata documents from 'deleteMetadata' methods

doulikecookiedough committed:
Add revert metadata deletion process to 'syncRenameMetadataDocForDeletion'

doulikecookiedough committed:
Add guard rails to 'syncRenameMetadataDocForDeletion' and new junit tests

doulikecookiedough committed:
Create new package 'hashstoreconverter' and add new class 'FileHashStoreLinks' with basic constructor

doulikecookiedough committed:
Add 'FileHashStoreLinks' constructor, new junit test class 'FileHashStoreLinksInitTest'

doulikecookiedough committed:
Add logging module to 'FileHashStoreLinks'

doulikecookiedough committed:
Change access modifier to public for 'enum DefaultHashAlgorithms' in 'FileHashStore', and add new method 'generateChecksums' in 'FileHashStoreLinks'

doulikecookiedough committed:
Rename 'FileHashStoreLinksInitTest' class to 'FileHashStoreLinks', add new junit test for 'generateChecksums'

doulikecookiedough committed:
Clean up code: remove unused imports

doulikecookiedough committed:
Add new method 'storeHardLink' to 'FileHashStoreLinks', new accessor method 'getHashStoreLinksDataObjectPath' and junit test for 'storeHardLink'

doulikecookiedough committed:
Clean up 'FileHashStoreLinks' class, remove unused variables

doulikecookiedough committed:
Add missing logging statement for 'storeHardLink'

doulikecookiedough committed:
Add missing class javadoc for 'FileHashStoreLinks'

doulikecookiedough committed:
Fix wording in javadoc

doulikecookiedough committed:
Add new HashStoreConverter class, public convert method and todo items

doulikecookiedough committed:
Clean-up code in 'HashStoreConverter' and add missing javadocs

doulikecookiedough committed:
Add new 'HashstoreConverter' test class

doulikecookiedough committed:
Catch and swallow 'FileAlreadyExistsException' in 'storeHardLink' and add new junit tests

doulikecookiedough committed:
Add new junit test for 'HashStoreConverter' convert method

doulikecookiedough committed:
Revise 'HashStoreConverter' javadocs to improve clarity

doulikecookiedough committed:
Update javadoc for 'HashStoreConverter'

doulikecookiedough committed:
Remove unused imports

doulikecookiedough committed:
Further revise javadocs

doulikecookiedough committed:
Cleanup and remove redundant junit tests, add missing javadocs

doulikecookiedough committed:
Refactor 'convert' method to conditionally store the data object or sysmeta, which may not always be available

doulikecookiedough committed:
Refactor 'convert' method to check for null sysmeta stream (it can never be null) and add junit test

doulikecookiedough committed:
Refactor 'convert' to store sysmeta first, which is a hard requirement, and add new junit tests

doulikecookiedough committed:
Add missing guard rail to check for null pid in 'convert' method

doulikecookiedough committed:
Revise 'convert' method by moving boolean directly into if statement

doulikecookiedough committed:
Extract new public method 'createParentDirectories' from 'FileHashStoreLinks' into 'FileHashStoreUtility'

doulikecookiedough committed:
Update HashStoreConverter's convert and FileHashStoreLinks' storeHardLink method signatures with 'checksum' and 'checksumAlgorithm', and revise junit tests

doulikecookiedough committed:
Clean-up 'convert' method and add new junit tests

doulikecookiedough committed:
Clean-up code for linting warnings, unused imports and redundant variables

doulikecookiedough committed:
Clean up javadocs

doulikecookiedough committed:
Revert 'convert' operation order back to storing hard link first before storing sysmeta

doulikecookiedough committed:
Update postgres version to 42.7.2 to resolve vulnerability warning

doulikecookiedough committed:
Update README.md for clarity

doulikecookiedough committed:
Rename 'deleteInvalidObject' method to 'deleteIfInvalidObject' and update affected code

doulikecookiedough committed:
Update .idea's .gitignore and 'README.md'

doulikecookiedough committed:
Revise javadoc usage of backticks to '{@code ... }'

doulikecookiedough committed:
Update maven and idea settings java target to 17 from 1.8

doulikecookiedough committed:
Add new junit test to check for when pid is simply a new line

doulikecookiedough committed:
Refactor FileHashStoreUtility 'checkNotNegativeOrZero' to 'checkPositive' and update affected code

doulikecookiedough committed:
Rename logging variable in 'HashStoreRunnable'

doulikecookiedough committed:
Clean up 'FileHashStoreUtility' logging and exception statements

doulikecookiedough committed:
Update maven.yml workflow to use JDK 17

doulikecookiedough committed:
Add missing javadocs in 'HashStoreRunnable' class and missing javadocs

doulikecookiedough committed:
Refactor 'FileHashStoreProtected' junit tests to create InputStreams in try statements for automatic resource management

doulikecookiedough committed:
Refactor 'FileHashStoreInit' junit tests to create InputStreams in try statements for automatic resource management

doulikecookiedough committed:
Refactor 'FileHashStoreClient' junit tests to create InputStreams in try statements for automatic resource management

doulikecookiedough committed:
Refactor 'HashStore' junit tests to create InputStreams in try statements for automatic resource management

doulikecookiedough committed:
Refactor 'FileHashStoreUtility' class' calculateHexDigest method to use try statement on the given stream for automatic resource management

doulikecookiedough committed:
Refactor 'FileHashStore' to use try statements for automatic resource management and 'buildHashStoreYamlString' as a text block for clarity

doulikecookiedough committed:
Refactor 'FileHashStoreInterface' junit tests to create InputStreams in try statements for automatic resource management where relevant

doulikecookiedough committed:
Add new enum 'HashStoreRefUpdateTypes' and refactor affected code in FileHashStore (related to updateRefsFile)

doulikecookiedough committed:
Refactor 'updateRefsFile' and 'getHashStoreRefsPath' to require enum parameter in signature, and refactor affected code and junit tests

doulikecookiedough committed:
Rename 'checkForEmptyAndValidString' method to 'checkForNotEmptyAndValidString' and update affected code

doulikecookiedough committed:
Apply autoformatting to 'syncPutObject'

doulikecookiedough committed:
Apply autoformatting to 'deleteMetadata'

doulikecookiedough committed:
Apply autoformatting to 'findObject'

doulikecookiedough committed:
Apply autoformatting to 'putObject'

doulikecookiedough committed:
Apply autoformatting to 'getHashStoreMetadataPath'

doulikecookiedough committed:
Remove redundant code in 'getHashStoreRefsPath' method

doulikecookiedough committed:
Add default case to 'getHashStoreRefsPath'

doulikecookiedough committed:
Refactor 'ObjectMetadata' to be a record instead of a custom class, and revise all junit tests and affected code

doulikecookiedough committed:
Cleanup 'ObjectMetadata' class

doulikecookiedough committed:
Add new record 'objectInfo', refactor 'findObject' to return an 'objectInfo' object and update junit tests

doulikecookiedough committed:
Refactor 'checkForNotEmptyAndValidString' to call '.isBlank()' instead of .trim() and then .isEmpty()

doulikecookiedough committed:
Refactor 'HashStoreRunnable' run's switch case per formatter suggestion

doulikecookiedough committed:
Update README.md

doulikecookiedough committed:
Update javadocs in 'HashStore' interface

doulikecookiedough committed:
Refactor and simplify usage of enum objects 'HashStoreIdTypes' and 'HashStoreRefUpdateTypes'

doulikecookiedough committed:
Add missing javadocs for enum objects to add clarity

doulikecookiedough committed:
Rename 'objectInfo' record to 'ObjectInfo'

doulikecookiedough committed:
Rename references of 'deleteInvalidObject' to 'deleteIfInvalidObject'

doulikecookiedough committed:
Refactor 'tagObject' by moving synchronization code to be closer to code requiring it in 'storeHashStoreRefsFiles'

doulikecookiedough committed:
Refactor 'checkForNotEmptyAndValidString' to get method name via thread and update signature to remove 'method' argument

doulikecookiedough committed:
Refactor 'ensureNotNull' to get method name via thread and update signature to remove 'method' argument

doulikecookiedough committed:
Refactor 'checkPositive' to get method name via thread and update signature to remove 'method' argument

doulikecookiedough committed:
Fix typo in 'HashStore' interface

doulikecookiedough committed:
Revise comments in 'tagObject'

doulikecookiedough committed:
Move synchronized call to within try statement in 'unTagObject' to improve flow

doulikecookiedough committed:
Apply IntelliJ automatic formatting to entire codebase for linting consistenty

doulikecookiedough committed:
Merge pull request #58 from DataONEorg/feature-57-deleteobj-refactor

Feature-57: `deleteObject` Refactor

doulikecookiedough committed:
Merge branch 'feature-57-deleteobj-refactor' into feature-95-hashstore-converter

doulikecookiedough committed:
Merge branch 'develop' into feature-95-hashstore-converter

doulikecookiedough committed:
Make enum 'DefaultHashAlgorithms' and 'HashStoreProperties' public to provide access to 'FileHashStoreLinks'

doulikecookiedough committed:
Revise 'FileHashStoreLinks' and 'HashStoreConverter' classes based on new changes from

doulikecookiedough committed:
Refactor 'storeHardLink' method to create InputStream from given filePath and update signature and affected junit tests

doulikecookiedough committed:
Refactor 'FileHashStore' and 'FileHashStoreLinks' to automate the creation and digestion of 'MessageDigest' objects based on the DefaultHashAlgorithms enum object

doulikecookiedough committed:
Optimize 'writeToTmpFileAndGenerateChecksums' and 'generateChecksums' methods

doulikecookiedough committed:
Make newly added list 'digestsToCalculate' private

doulikecookiedough committed:
Further optimize code to generate checksums by using try-resources on InputStream object

doulikecookiedough committed:
Fix bug in '...Checksums' optimization by creating a local list and modifying it, rather than using a class variable and accessing it directly

doulikecookiedough committed:
Revise and add new junit tests related to '...Checksums' methods

doulikecookiedough committed:
Change previously public access modifiers for FileHashStore class variables to protected

doulikecookiedough committed:
Rename 'defaultHexDigestsList' to 'defaultMessageDigestsList' and revise comments

doulikecookiedough committed:
Merge pull request #96 from DataONEorg/feature-95-hashstore-converter

Feature-95: HashStoreConverter & FileHashStoreLinks

doulikecookiedough committed:
Remove unused import in 'HashStoreConverter'

doulikecookiedough committed:
Update junit test debug messaging for better accuracy (.getMessage instead of .getCause)

doulikecookiedough committed:
Roll back optimizations to the default hash algorithm calculation process

doulikecookiedough committed:
Re-apply IntelliJ auto formatting

doulikecookiedough committed:
Merge pull request #99 from DataONEorg/bug-98-storeobj-failedtest

Bug-98: `storeObject` Failed Junit Test

doulikecookiedough committed:
Declare 'NonMatchingChecksumException' in HashStoreConverter's 'convert' method and add new junit test

doulikecookiedough committed:
Fix inaccurate javadoc for new junit test

doulikecookiedough committed:
Cleanup 'FileHashStoreLinksTest' class

doulikecookiedough committed:
Merge pull request #102 from DataONEorg/feature-101-nmce-declare

Feature-101: 'NonMatchingChecksumException' Declaration

doulikecookiedough committed:
Remove 'unTagObject' call from 'tagObject' method

doulikecookiedough committed:
Refactor 'storeHashStoreRefsFiles' to 'unTagObject'

doulikecookiedough committed:
Add new custom exception 'IdentifierNotLockedException'

doulikecookiedough committed:
Refactor 'unTagObject' with guard rails to ensure thread safety, revise access modifier for sync methods to be protected to assist with testing

doulikecookiedough committed:
Refactor junit tests for 'unTagObject' after making changes

doulikecookiedough committed:
Cleanup code, comments and javadocs

doulikecookiedough committed:
Re-add 'static' access modifiers to synchronized methods

taojing2002 committed:
Added the code to deploy the jar files to the remote maven repostiory.

taojing2002 committed:
Added two profile to specify the deployment location: local or remote.

taojing2002 committed:
Merge pull request #105 from DataONEorg/feature-104-deploy-repository

Feature 104 deploy repository

doulikecookiedough committed:
Fix bug in 'storeHashStoreRefsFiles' where we could unintentionally overwrite a pid reference file and update junit test

doulikecookiedough committed:
Refactor 'storeHashStoreRefsFiles' to always throw exception if a pid refs file already exists

doulikecookiedough committed:
Refactor 'storeHashStoreRefsFiles' exception scenario to include more context

doulikecookiedough committed:
Refactor 'updateRefsFile' so that duplicate values are never added and add new junit test

doulikecookiedough committed:
Adjust 'storeHashStoreRefs' scenario when pid refs doesn't exist but cid refs does to create pid refs file first

doulikecookiedough committed:
Clean up/optimize 'updateRefsFile' by combining if statements

doulikecookiedough committed:
Refactor 'unTagObject' to proceed with as much of the untagging flow as possible by wrapping relevant code blocks in try-catch statements

doulikecookiedough committed:
Clean up 'unTagObject' and re-order process to focus on pid-related tasks

doulikecookiedough committed:
Further clean-up 'unTagObject' to proceed with as much of the untagging flow

doulikecookiedough committed:
Revise exception messaging to improve clarity in 'unTagObject'

doulikecookiedough committed:
Add missing javadoc exception in 'findObject'

doulikecookiedough committed:
Revise 'unTagObject' to check that cid retrieved equals to the cid from the call

doulikecookiedough committed:
Revise 'unTagObject' OrphanRefsFileException scenario to check cid before proceeding with untagging, and add new junit tests

doulikecookiedough committed:
Add additional guard rails for when retrieving/reading cid values, and ensuring cid values are locked

doulikecookiedough committed:
Add missing guard rail in 'unTagObjet' PidRefsFileNotFoundException for confirming cid is locked

doulikecookiedough committed:
Improve exception messaging in 'unTagObject'

doulikecookiedough committed:
Revise exception type in 'unTagObject' when cid doesn't match, add todo items

doulikecookiedough committed:
Update and add a new junit for revised exceptions

doulikecookiedough committed:
Refactor 'unTagObject' by extracting method 'addAndRenamePidRefsFileToDeleteList'

doulikecookiedough committed:
Refactor 'unTagObject' by extracting method 'deleteListOfFilesRenamedForDeletion'

doulikecookiedough committed:
Refactor 'unTagObject' by extracting method 'removePidFromCidRefsAndDetermineDeletion'

doulikecookiedough committed:
Refactor 'unTagObject' by extracting method 'validateCidAndCheckLocked'

doulikecookiedough committed:
Refactor 'unTagObject' to always confirm cid before proceeding and resolve todo items

doulikecookiedough committed:
Refactor 'unTagObject' further by extracting relevant code to 'removePidFromCidRefsAndDetermineDeletion'

doulikecookiedough committed:
Rename 'removePidFromCidRefsAndDetermineDeletion' method to 'removePidAndHandleCidDeletion'

doulikecookiedough committed:
Rename 'deleteListOfFilesRenamedForDeletion' method to 'deleteMarkedFiles'

doulikecookiedough committed:
Rename 'addAndRenamePidRefsFileToDeleteList' method to 'markPidRefsFileForDeletion'

doulikecookiedough committed:
Rename 'validateCidAndCheckLocked' method to 'validateAndCheckCidLock'

doulikecookiedough committed:
Change warning log levels to error level for 'unTagObject' related methods

doulikecookiedough committed:
Further revise logging levels for 'unTagObject'

doulikecookiedough committed:
Revise inaccurate variable name

doulikecookiedough committed:
Merge pull request #103 from DataONEorg/bug-97-tagobject-orphan

Bug-97: `tagObject` Produces Orphaned Object

doulikecookiedough committed:
Update 'pom.xml' hashstore-java version to 1.1.0

doulikecookiedough committed:
Refactor 'generateTmpFile' in 'FileHashStoreUtility' to set 'rw-r-----' file permissions on tmp file and add new junit test

doulikecookiedough committed:
Add new junit test for 'storeObject' to confirm that file permissions persist after moving a tmp file to its permanent location

doulikecookiedough committed:
Rename junit test to match method being tested

doulikecookiedough committed:
Revise new junit tests for tmpFile permissions to be more thorough (check for absence of other permissions)

doulikecookiedough committed:
Add 'final' to permissions variable in 'generateTmpFile' for clarity

doulikecookiedough committed:
Revise javadoc for 'generateTmpFile' and fix minor formatting issue with new junit tests

doulikecookiedough committed:
Merge pull request #109 from DataONEorg/feature-108-tmpfile-permissions

Feature-108: Default Permissions on HashStore Objects

doulikecookiedough committed:
Revise 'pom.xml' hashstore-java version to '1.1.0-SNAPSHOT'

doulikecookiedough committed:
Revise 'createParentDirectories' method in 'FileHashStoreUtility' to set 'rwxr-x---' permissions

doulikecookiedough committed:
Refactor 'FileHashStore' move method to use 'FileHashStoreUtility.createParentDirectories' and add new junit tests

doulikecookiedough committed:
Change logging to debug from warn in 'FileHashStore' and 'FileHashStoreLinks' when duplicate object has been encountered

doulikecookiedough committed:
Remove unused imports

doulikecookiedough committed:
Merge pull request #112 from DataONEorg/feature-111-directory-permissions

Feature-111: Directory Permissions

doulikecookiedough committed:
Add human-ness to 'CONTRIBUTING.md'

doulikecookiedough committed:
Revise 'README.md' for accuracy and client example

doulikecookiedough committed:
Merge pull request #113 from DataONEorg/feature-106-relpep-part1

Feature-106: Release Prep Part 1

doulikecookiedough committed:
Revise 'NonMatchingChecksumException' class to have a 'hexDigests' attribute

doulikecookiedough committed:
Include hexDigests when 'NonMatchingChecksumException' is thrown in 'FileHashStore' and 'FileHashStoreLinks'

doulikecookiedough committed:
Add new junit test to confirm hexDigests are included when 'NonMatchingChecksumException' is thrown and fix innacurate javadoc for a junit test

doulikecookiedough committed:
Add null check for 'validateTmpObject' method and new junit test

doulikecookiedough committed:
Add null check for hexDigests in 'storeHardLink' method for 'FileHashStoreLinks' to be safe

doulikecookiedough committed:
Merge pull request #115 from DataONEorg/feature-114-nmce-includechecksums

Feature-114: NonMatchingChecksumException hexDigests Inclusion

doulikecookiedough committed:
Revise junit test for 'NonMatchingCheckException' to also check that an additional algorithm is included if supplied, and add new junit test for 'storeHardLink'

doulikecookiedough committed:
Update 'pom.xml' version to 1.1.0 to prepare for release

doulikecookiedough committed:
Update 'README.md' with citation section and version number
